Networking: Add Fees to Orders

ctarda opened the pull request:
This is the first step to add support for Fees to the Order Details page. (#3084)

This PR implements support for Fees in Networking. Support for Fees in Yosemite and Storage will be added in subsequent PRs.

## Changes
- Add new model objects to Networking: `OrderFeeLine`, `OrderFeeTaxStatus` 
- Update `Order` to add a new property called `fees`
- Update mocks and json payloads in tests target
- Add a test to OrderMapperTests to test the parsing of an individual Order

## How to test
There are no user facing changes. 
- 👀 at the code 
- Check for a ✅ on CI.

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ctarda committed:
Add Fees to Order in Networking

- Includes updated mocks
- Includes tests

jaclync commented on Networking/Networking/Model/OrderFeeLine.swift:
```diff
@@ -0,0 +1,69 @@
+// MARK: - Comparable Conformance
+//
+extension OrderFeeLine: Comparable {
+    public static func == (lhs: OrderFeeLine, rhs: OrderFeeLine) -> Bool {
+        return lhs.feeID == rhs.feeID &&
+            lhs.name == rhs.name &&
+            lhs.total == rhs.total &&
+            lhs.totalTax == rhs.totalTax
+    }
```
nit: do we intentionally choose this custom `Equatable` implementation, or does it make sense to conform to `Equatable` with the default implementation which compares all attributes? In other models, we generally use the default `Equatable` implementation so I was just curious when we choose to compare just a subset of attributes

jaclync commented on Networking/Networking/Model/OrderFeeLine.swift:
```diff
@@ -0,0 +1,69 @@
+            lhs.totalTax == rhs.totalTax
+    }
+
+    public static func < (lhs: OrderFeeLine, rhs: OrderFeeLine) -> Bool {
+        return lhs.feeID < rhs.feeID ||
+            (lhs.feeID == rhs.feeID && lhs.name < rhs.name) ||
+            (lhs.feeID == rhs.feeID && lhs.name == rhs.name && lhs.total < rhs.total)
+    }
```
nit: do we need this for the use case?

jaclync commented on Networking/NetworkingTests/Mapper/OrderMapperTests.swift:
```diff
@@ -254,6 +254,27 @@ final class OrderMapperTests: XCTestCase {
+        XCTAssertEqual(fee.name, "$125.50 fee")
+        XCTAssertEqual(fee.taxStatus, .taxable)
+        XCTAssertEqual(fee.total, "125.50")
+        XCTAssertEqual(fee.totalTax, "0.00")
```
nit: do we also want to compare other fields like `taxClass` and `taxes`? or compare two objects if we use the default `Equatable` implementation?

ctarda replied:
🤔 good point. It probably makes more sense to just rely on the default implementations of Equatable and compare two objects directly. 

Thanks @jaclync ! I'll update and fix the breaking test

jaclync reviewed (commented):
Thanks for taking on this task adding fees!

Code: just some minor questions about the custom `Equatable` implementation and comparison!
Testing: `OrderStoreTests_FetchFilteredAndAllOrders.testWhenGivenAFilterItFetchesBothTheFilteredListAndTheAllOrdersList` didn't pass in CI, probably from the orders in `orders-load-all-2` not all having `fee_lines` field and thus these 4 orders can’t be parsed?

ctarda committed:
Remove custom implementation of Equatable and Comparable

ctarda committed:
Fix breaking test

ctarda committed:
Touch all properties of OrderFeeLine in test

ctarda commented:
This is ready for another round @jaclync 

I removed the custom implementation of Equatable and Comparable, updated the mock so all tests should pass (sorry about that, 😬 ) and in the end went with touching all the object properties in the test, as it seemed to make that particular test a bit more explicit than testing equality, and it blended a bit better with the tests around it. Please let me know what you think

ctarda committed:
Merge branch 'develop' into issue/3084-order-fees-marki-networking

ctarda committed:
Add back fee_lines to parameters of OrdersRemote

jaclync reviewed (approved):
LGTM! :shipit:
